File: competitive-programming/src/edu/algos/li18_graph_grid/easy/Dfs_L733_FloodFill.java
package edu.algos.li18_graph_grid.easy;

import java.util.Arrays;

public class Dfs_L733_FloodFill {

	public static void main(String[] args) {
		Dfs_L733_FloodFill obj = new Dfs_L733_FloodFill();
		
		int[][] image = new int[][]{{1,1,1},{1,1,0},{1,0,1}};
		int[][] newImage = obj.floodFill(image, 1, 1, 2);
		for(int[] row: newImage) {
			System.out.println(Arrays.toString(row));
		}
	}

	/*
	 * Problem: https://leetcode.com/problems/flood-fill/
	 * Idea: Do dfs recursively in 4 directions and pass first src cell color and new color
     * If the color of current cell is same as src color, change it to new color.
	 * */
	public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(color == image[sr][sc]) return image;
        dfs(image, sr, sc, image[sr][sc], color);
        return image;
    }
    
    private void dfs(int[][] image, int i, int j, int srcColor, int newColor) {
        if(isSafe(image, i, j, srcColor)){
            image[i][j] = newColor;
            dfs(image, i+1, j, srcColor, newColor);
            dfs(image, i-1, j, srcColor, newColor);
            dfs(image, i, j+1, srcColor, newColor);
            dfs(image, i, j-1, srcColor, newColor);
        }
    }
    
    private boolean isSafe(int[][] image, int i, int j, int color) {
        int m = image.length, n = image[0].length;
        
        return i>=0 && i<m && j>=0 && j<n && color == image[i][j];
    }
}

File: dsa/topics/13_graphs/python/dfs_l733_flood_fill.py
# Problem:
# https://leetcode.com/problems/flood-fill/
#
# Idea:
# - Perform DFS from the starting cell (sr, sc)
# - Store the original source color
# - Recursively move in 4 directions
# - If the current cell color matches the source color,
#   repaint it with the new color
# - Stop recursion when out of bounds or color doesn't match
#
# Time: O(m * n)
# Space: O(m * n)  (recursion stack)


def flood_fill(image, sr, sc, color):
    rows, cols = len(image), len(image[0])
    src_color = image[sr][sc]

    # If source color and new color are same, no change needed
    if src_color == color:
        return image

    # DFS to repaint connected pixels
    def dfs(r, c):
        # Check if cell is inside grid and matches source color
        if not (0 <= r < rows and 0 <= c < cols):
            return
        if image[r][c] != src_color:
            return

        image[r][c] = color

        for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
            dfs(r + dr, c + dc)

    dfs(sr, sc)
    return image


# -------------------- TEST --------------------
if __name__ == "__main__":
    image = [
        [1, 1, 1],
        [1, 1, 0],
        [1, 0, 1]
    ]

    result = flood_fill(image, 1, 1, 2)
    for row in result:
        print(row)
